---
title: "Testing CRISPR preprocessing with pseudoalignment"
author: Candace Savonen
date: 2024
output:   
  html_notebook:
    toc: true
    toc_float: true
---

## Objectives
---

```bash
salmon index -t config/ref/pgPEN_R1.fa -i r1_index --keepDuplicates -k 9
salmon index -t config/ref/pgPEN_R2.fa -i r2_index --keepDuplicates -k 9
```

## Quantification with Salmon

```bash
salmon quant -i transcripts_for_index \
    -l A \
    -r data/fastqs/fastq_demuxed/pgMAP_tutorial_gRNA1_trimmed_sample1.fastq.gz \
    -o data/salmon_quant/r1_index \
    --validateMappings \
    --threads 4


salmon quant -i transcripts_rev_index \
    -l A \
    -r data/fastqs/fastq_demuxed/pgMAP_tutorial_gRNA2_trimmed_sample1.fastq.gz \
    -o data/salmon_quant/r2_index \
    --validateMappings \
    --threads 4
```


```{r}
# This is using already aligned reads but salmon to quantify.
salmon quant -t config/ref/pgPEN_R1.fa -l A -a data/aligned_bam/sample1_forward_file.bam -o salmon_quant_aligned/salmon_quant_align_1
salmon quant -t config/ref/pgPEN_R2.fa -l A -a data/aligned_bam/sample1_reverse_file.bam -o salmon_quant_aligned/salmon_quant_align_2
```

## Libraries and functions

```{r library}
# Load magrittr for the pipe
library(magrittr)
```

## Directories and files

```{r directories, live = TRUE}
# directory where the data are located
data_dir <- file.path("..", "data")

# directory where the quant files are located, each sample is its own
# directory
quant_dir <- file.path(data_dir, "salmon_quant")

# create a directory to hold the tximeta results if it doesn't exist yet
txi_dir <- file.path(data_dir, "txi")
if (!dir.exists(txi_dir)) {
  dir.create(txi_dir, recursive = TRUE)
}
```

We'll need the `quant.sf` files for all the samples in an experiment which we have stored in `quant_dir`.

```{r input-names}
# the quant files themselves
sf_files <- list.files(quant_dir, recursive = TRUE, full.names = TRUE,
                       pattern = "quant.sf")
```

**Output**

```{r output-names, live = TRUE}
# Name the output gastric-cancer_tximeta.RDS and use the directory created
# above as the rest of the path
txi_out_file <- file.path(txi_dir, "tximeta.RDS")
```

## File names

All output files from `salmon quant` we'll use with `tximeta` are named `quant.sf`.
Unfortunately, this means that the file names themselves do not have any information about the sample they come from!

```{r sf_files, live = TRUE}
# Let's look at the full path for the quant.sf files
sf_files
```

Let's extract the _sample_ names from the **file paths** using the `stringr` package.

Notice how the file path is separated by `/`.
If we were to split up this character string by `/`, the second to last item is the sample names (because we used them as directory names for the `salmon` output).
This is exactly what `stringr::word()` allows us to do: split up the file paths by `/` and extract the sample names.

```{r sample_names}
sample_names <- stringr::word(sf_files, -2, sep = "/")
sample_names
```

## Set up metadata

`tximeta` needs a data frame with at least these two columns:
- a `files` column  with the file paths to the quant.sf files
- a `names` column with the sample names

```{r names_sf_files}
coldata <- data.frame(files = sf_files,
                      names = c("sample1_r1", "sample1_r2"))
```

We have more information about these samples stored in the metadata file that we will also want stored in `coldata`.


```{r}
quant_for <- readr::read_tsv(sf_files[1])
quant_rev <- readr::read_tsv(sf_files[2])

quant <- dplyr::inner_join(quant_for, quant_rev, by = "Name",
                           suffix = c("_r1", "_r2")) %>%
  dplyr::filter(NumReads_r1 != 0,
                NumReads_r2 != 0)


quant$minRead <- apply(quant[,c(5,9)], 1, min)
quant$meanRead <- apply(quant[,c(5,9)], 1, mean)
```

```{r}
original_counts <- readr::read_csv("counts_pgmap.tsv") %>%
  dplyr::filter(!is.na(sample1))
```


```{r}
quant_all <- quant %>%
  dplyr::left_join(original_counts, by = c("Name" = "id"))


plot(quant_all$minRead, quant_all$sample1)
plot(quant_all$TPM_r1, quant_all$sample1)

quant_all$minRead - quant_all$sample1
```

```{r}
plot(quant_all$meanRead, quant_all$sample1)
```

```{r}
cor(quant_all$minRead, quant_all$sample1)
```
## Session Info

Record session info for reproducibility & provenance purposes.

```{r sessioninfo}
sessionInfo()
```
