---
title: "Getting Started"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{getting-started}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# pgmap tutorial


## Build Indices

```{r}
# Declare an input file path for FORWARD
for_ref <- file.path(config_folder(), "ref", "paralog_pgRNA1.fa")
# Declare ann output file path for FORWARD
for_index_file_path <- file.path(config_folder(), "pgPEN_index_for")
# Declare an input file path for REVERSE
rev_ref <- file.path(config_folder(), "ref", "paralog_pgRNA2.fa")
# Declare a output file path for REVERSE
rev_index_file_path <- file.path(config_folder(), "pgPEN_index_rev")
```

```{r}
 # Build the FORWARD reference
 bowtie2_build(
   references = for_ref,
   bt2Index = for_index_file_path,
   overwrite = TRUE
 )
 # Build the REVERSE reference
 bowtie2_build(
   references = rev_ref,
   bt2Index = rev_index_file_path,
   overwrite = TRUE
 )
```

## Declare file paths

```{r}
# These MUST be names that are listed in the bam file name itself.
# There needs to be exactly 2 bam files per sample
sample_names <- c("sample1", "sample2", "sample3")

# This will be the directory that contains all the file path
fastq_dir <- file.path(example_data_folder(), "fastqs", "fastq_demuxed")

# This will be the directory that contains all the file path
bam_dir <- file.path(example_data_folder(), "bam")
```

## Demux potentially here

```{r}
## TODO: 
```

## Quality control

```{r}
fastqc(fq.dir = fastq_dir, # FASTQ files directory
       qc.dir = , # Results directory
       threads = 4                    # Number of threads
       )
```

## Alignment

```{r}
fastq_to_bam(
 fastq_dir = fastq_dir,
 for_index = for_index_file_path,
 rev_index = rev_index_file_path,
 sample_names = sample_names,
 output_dir =  bam_dir,
 time = TRUE,
 overwrite = TRUE
)
```

## Get stats

```{r}
 counts_df <- calc_counts(
   bam_dir = bam_dir,
   sample_names = sample_names,
   time = TRUE
 )
```

## Calculate stats

```{r}
stats_df <- get_stats(
  bam_dir = bam_dir,
  sample_names = sample_names,
  output_dir = file.path(example_data_folder(), "bam_stats")
)
```

## Write to files

```{r}
 # We can write these to CSV files like this
 readr::write_csv(counts_df$counts, "counts_pgmap.tsv")
 readr::write_csv(counts_df$stats, "stats_pgmap.tsv")
```

## Session Info

This is just for provenance purposes.

```{r}
sessionInfo()
```
