<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>getting_started • gimap</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="getting_started">
<meta property="og:description" content="gimap">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">gimap</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/getting_started.html">getting_started</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/FredHutch/gimap/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>getting_started</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/FredHutch/gimap/blob/HEAD/vignettes/getting_started.Rmd" class="external-link"><code>vignettes/getting_started.Rmd</code></a></small>
      <div class="hidden name"><code>getting_started.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="gimap-tutorial">gimap tutorial<a class="anchor" aria-label="anchor" href="#gimap-tutorial"></a>
</h2>
<p>gimap performs analysis of dual-targeting CRISPR screening data, with
the goal of aiding the identification of genetic interactions
(e.g. cooperativity, synthetic lethality) in models of disease and other
biological contexts. gimap analyzes functional genomic data generated by
the pgPEN (paired guide RNAs for genetic interaction mapping) approach,
quantifying growth effects of single and paired gene knockouts upon
application of a CRISPR library. A multitude of CRISPR screen types can
be used for this analysis, with helpful descriptions found in this
review (<a href="https://www.nature.com/articles/s43586-021-00093-4" class="external-link uri">https://www.nature.com/articles/s43586-021-00093-4</a>). Use
of pgPEN and GI-mapping in a paired gRNA format can be found here (<a href="https://pubmed.ncbi.nlm.nih.gov/34469736/" class="external-link uri">https://pubmed.ncbi.nlm.nih.gov/34469736/</a>).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/FredHutch/gimap" class="external-link">gimap</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span></code></pre></div>
<div class="section level3">
<h3 id="data-requirements">Data requirements<a class="anchor" aria-label="anchor" href="#data-requirements"></a>
</h3>
<p>Let’s examine this example pgPEN counts table. It’s divided into
columns containing:</p>
<ul>
<li>
<code>id</code>: an ID corresponding to the names of paired
guides</li>
<li>
<code>seq_1</code>: gRNA sequence 1, targeting “paralog A”</li>
<li>
<code>seq_2</code>: gRNA sequence 2, targeting “paralog B”</li>
<li>
<code>Day00_RepA</code>: Gene Counts from Day 00 for Replicate
A</li>
<li>
<code>Day05_RepA</code>: Gene Counts from Day 05 for Replicate
A</li>
<li>
<code>Day22_RepA</code>: Gene Counts from Day 22 for Replicate
A</li>
<li>
<code>Day22_RepB</code>: Gene Counts from Day 22 for Replicate
B</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_example_data.html">get_example_data</a></span><span class="op">(</span><span class="st">"count"</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">33170</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">8</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Delimiter:</span> "\t"</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">chr</span> (3): id, seq_1, seq_2</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">dbl</span> (5): Day00_RepA, Day05_RepA, Day22_RepA, Day22_RepB, Day22_RepC</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.</span></span></code></pre></div>
<p>The metadata you have may vary slightly from this but you’ll want to
make sure you have the essential variables and information regarding how
you collected your data.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">example_data</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "id"         "seq_1"      "seq_2"      "Day00_RepA" "Day05_RepA"</span></span>
<span><span class="co">#&gt; [6] "Day22_RepA" "Day22_RepB" "Day22_RepC"</span></span></code></pre></div>
<p>This count data has …</p>
<p>We also want the construct metadata</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_pg_metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_example_data.html">get_example_data</a></span><span class="op">(</span><span class="st">"meta"</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">33170</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">9</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Delimiter:</span> ","</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">chr</span> (9): pgRNA_ID, target1_sgRNA_seq, target2_sgRNA_seq, paralog_pair, targe...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">example_pg_metadata</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "pgRNA_ID"           "target1_sgRNA_seq"  "target2_sgRNA_seq" </span></span>
<span><span class="co">#&gt; [4] "paralog_pair"       "target_type"        "target1"           </span></span>
<span><span class="co">#&gt; [7] "target2"            "target1_ensembl_id" "target2_ensembl_id"</span></span></code></pre></div>
<p>And the meta data has …</p>
</div>
<div class="section level3">
<h3 id="setting-up-data">Setting up data<a class="anchor" aria-label="anchor" href="#setting-up-data"></a>
</h3>
<p>We’re going to set up three datasets. Two are required; the first is
the counts that the genetic interaction analysis will be used for.</p>
<p>The first data set contains the read counts from each sample type.
Required for analysis is a Day 0 (or plasmid) sample, and at least one
further timepoint sample. The T0 sample, or plasmid sample, will
represent the entire library before any type of selection has occurred
during the length of the screen. This is the baseline for guide RNA
representation. The length of time cells should remain in culture
throughout the screen is heavily dependent on the type of selection
occurring, helpful advice can be found in (<a href="https://www.nature.com/articles/s43586-021-00093-4" class="external-link uri">https://www.nature.com/articles/s43586-021-00093-4</a>). QC
analysis will follow to correlate replicates if inputted. Comparison of
early and late timepoints is possible in this function, but not required
if early timepoints were not taken.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_counts</span> <span class="op">&lt;-</span> <span class="va">example_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Day00_RepA"</span>, <span class="st">"Day05_RepA"</span>, <span class="st">"Day22_RepA"</span>, <span class="st">"Day22_RepB"</span>, <span class="st">"Day22_RepC"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>The next datasets are metadata that describe the dimensions of the
count data. - These both need to be data frames. - The sizes of these
metadata must correspond to the dimensions of the counts data.</p>
<p><code>pg_id</code> are just the unique IDs listed in the same
order/sorted the same way as the count data and can be used for mapping
between the count data and the metadata.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_pg_id</span> <span class="op">&lt;-</span> <span class="va">example_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="st">"id"</span><span class="op">)</span></span></code></pre></div>
<p>One of these (<code>example_pg_metadata</code>) is required because
it is necessary to know the IDs and be able to map them to pgRNA
constructs.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_pg_metadata</span> <span class="op">&lt;-</span> <span class="va">example_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"id"</span>, <span class="st">"seq_1"</span>, <span class="st">"seq_2"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Sample metadata is the information that describes the samples and is
sorted the same way as the columns in the count data.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_sample_metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  id <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,</span>
<span>  day <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Day00"</span>, <span class="st">"Day05"</span>, <span class="st">"Day22"</span>, <span class="st">"Day22"</span>, <span class="st">"Day22"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  rep <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RepA"</span>, <span class="st">"RepA"</span>, <span class="st">"RepA"</span>, <span class="st">"RepB"</span>, <span class="st">"RepC"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We’ll need to provide <code>example_counts</code>,
<code>pg_ids</code> and <code>pg_metadata</code> to
<code><a href="../reference/setup_data.html">setup_data()</a></code>. We can provide <code>sample_metadata</code>,
but it is not required at the moment.</p>
<p>Now let’s setup our data using <code><a href="../reference/setup_data.html">setup_data()</a></code>. Optionally
we can provide the metadata in this function as well so that it is
stored with the data.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gimap_dataset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/setup_data.html">setup_data</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">example_counts</span>,</span>
<span>                            pg_ids <span class="op">=</span> <span class="va">example_pg_id</span>,</span>
<span>                            pg_metadata <span class="op">=</span> <span class="va">example_pg_metadata</span>,</span>
<span>                            sample_metadata <span class="op">=</span> <span class="va">example_sample_metadata</span><span class="op">)</span></span></code></pre></div>
<p>You’ll notice that this set up gives us a list of formatted data.
This contains the original counts we gave <code><a href="../reference/setup_data.html">setup_data()</a></code>
function but also normalized counts, and the total counts per
sample.</p>
<ul>
<li>
<code>raw_counts</code>: The original counts data that illustrates
the number of cells alive in the sample. This data has samples as the
columns and the paired guide constructs as rows.</li>
<li>
<code>counts_per_sample</code>: Add up all the counts for each
sample over all of the paired guide designs.</li>
<li>Transformed data: This section contains the various types of
normalized and adjusted data made from the raw counts data.</li>
<li>
<code>count_norm</code> - For each sample, the data is normalized
<code>-log10(( counts +1) /  total counts for the sample over all the pg designs ))</code>
</li>
<li>
<code>cpm</code> - For each sample this is calculated by taking the
<code>counts / total counts for the sample over all the pg designs)*1e6</code>
</li>
<li>
<code>log2cpm</code>: log-2 transformed counts per million this is
calculated by <code>log2(cpms + 1)</code>
</li>
<li>pg_metadata: paired guide metadata - information that describes the
paired-guided RNA designs. This may include the sequences used in the
CRISPR design as well as what genes are targeted.</li>
<li>
<code>sample_metadata</code>: Metadata that describes the samples.
This likely includes the time point information, replicates, sample IDs,
and any other additional information that is needed regarding the
experimental setup.</li>
</ul>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">gimap_dataset</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 8</span></span>
<span><span class="co">#&gt;  $ raw_counts       : num [1:33170, 1:5] 951 1511 191 588 289 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ : NULL</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:5] "Day00_RepA" "Day05_RepA" "Day22_RepA" "Day22_RepB" ...</span></span>
<span><span class="co">#&gt;  $ counts_per_sample: Named num [1:5] 38471205 32028290 29144511 36628141 36448291</span></span>
<span><span class="co">#&gt;   ..- attr(*, "names")= chr [1:5] "Day00_RepA" "Day05_RepA" "Day22_RepA" "Day22_RepB" ...</span></span>
<span><span class="co">#&gt;  $ transformed_data :List of 3</span></span>
<span><span class="co">#&gt;   ..$ count_norm: num [1:33170, 1:5] 4.61 4.41 5.3 4.82 5.12 ...</span></span>
<span><span class="co">#&gt;   .. ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. .. ..$ : NULL</span></span>
<span><span class="co">#&gt;   .. .. ..$ : chr [1:5] "Day00_RepA" "Day05_RepA" "Day22_RepA" "Day22_RepB" ...</span></span>
<span><span class="co">#&gt;   ..$ cpm       : num [1:33170, 1:5] 24.72 39.28 4.96 15.28 7.51 ...</span></span>
<span><span class="co">#&gt;   .. ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. .. ..$ : NULL</span></span>
<span><span class="co">#&gt;   .. .. ..$ : chr [1:5] "Day00_RepA" "Day05_RepA" "Day22_RepA" "Day22_RepB" ...</span></span>
<span><span class="co">#&gt;   ..$ log2_cpm  : num [1:33170, 1:5] 4.68 5.33 2.58 4.03 3.09 ...</span></span>
<span><span class="co">#&gt;   .. ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. .. ..$ : NULL</span></span>
<span><span class="co">#&gt;   .. .. ..$ : chr [1:5] "Day00_RepA" "Day05_RepA" "Day22_RepA" "Day22_RepB" ...</span></span>
<span><span class="co">#&gt;  $ metadata         :List of 3</span></span>
<span><span class="co">#&gt;   ..$ pg_ids         : tibble [33,170 × 1] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co">#&gt;   .. ..$ id: chr [1:33170] "AADAC_AADACL2_pg1" "AADAC_AADACL2_pg10" "AADAC_AADACL2_pg11" "AADAC_AADACL2_pg12" ...</span></span>
<span><span class="co">#&gt;   ..$ pg_metadata    : NULL</span></span>
<span><span class="co">#&gt;   ..$ sample_metadata:'data.frame':  5 obs. of  3 variables:</span></span>
<span><span class="co">#&gt;   .. ..$ id : int [1:5] 1 2 3 4 5</span></span>
<span><span class="co">#&gt;   .. ..$ day: Factor w/ 3 levels "Day00","Day05",..: 1 2 3 3 3</span></span>
<span><span class="co">#&gt;   .. ..$ rep: Factor w/ 3 levels "RepA","RepB",..: 1 1 1 2 3</span></span>
<span><span class="co">#&gt;  $ filtered_data    : NULL</span></span>
<span><span class="co">#&gt;  $ annotation       : NULL</span></span>
<span><span class="co">#&gt;  $ log_fc           : NULL</span></span>
<span><span class="co">#&gt;  $ results          : NULL</span></span>
<span><span class="co">#&gt;  - attr(*, "class")= chr [1:2] "list" "gimap_dataset"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="quality-checks">Quality Checks<a class="anchor" aria-label="anchor" href="#quality-checks"></a>
</h3>
<p>The first step is running some quality checks on our data. The
<code><a href="../reference/run_qc.html">run_qc()</a></code> function will create a report we can look at to
assess this.</p>
<p>The report includes several visualizations of raw/unfiltered
data:</p>
<ul>
<li>distribution of normalized counts for each sample. The goal of
determining this distribution is to identify pgRNA counts that are low
at the start of the screen - before any type of phenotypic or growth
selection is occurring, either in the T0 or plasmid sample. These low
abundance pgRNAs should be removed from the analysis. [the goal section
here doesn’t fit my expectations/understanding of this plot.]</li>
<li>histogram of <code>log2cpm</code> values for each individual sample:
this helps users identify samples that do not have a normal distribution
of reads and inform the upcoming filtering steps.</li>
<li>sample correlation heatmap: generates a heatmap using cpm values for
each replicate/sample. This heatmap gives an overview on how similar
samples are. Replicates should correlate well, and cluster together,
while each timepoint sample should be different from the T0. This
analysis will also allow users to identify potential sample swaps, if
correlation scores between replicates is poor.</li>
<li>A histogram that shows the variance within replicates for each pgRNA
construct. For each pgRNA construct, the variance among the 3 replicates
is found and a distribution is constructed by looking at these variances
together.</li>
<li>A histogram of the log2 CPM values of pgRNA constructs at the
plasmid time point. This graph relates to a plasmid filter that can be
applied to the data, but the plot displays all of the data prior to a
filter being applied.</li>
</ul>
<p>This report also includes several visualizations after filters are
applied:</p>
<p>There is a filter that flags pgRNA constructs where any of the time
points have a count of zero. - We include a bar plot that shows the
number of pgRNA constructs which have counts of zero in either 0, 1, 2,
or 3 replicates. - We include a table that specifies how many pgRNAs
would be filtered out by applying this filter.</p>
<p>There is a filter that flags pgRNA constructs that have low log2 CPM
counts for the day 0 or plasmid time point. - The histogram of the log2
CPM values of pgRNA constructs at the plasmid time point mentioned
earlier does have a dashed line specifying the lower outlier (or a user
defined cutoff) and pgRNA constructs with a plasmid log2 CPM lower than
that value can be filtered out. - We include a table that specifies how
many pgRNAs would be filtered out by applying this filter.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/run_qc.html">run_qc</a></span><span class="op">(</span><span class="va">gimap_dataset</span>,</span>
<span>       output_file <span class="op">=</span> <span class="st">"example_qc_report.Rmd"</span>,</span>
<span>       overwrite <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>       quiet <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Results in: example_qc_report.html</span></span></code></pre></div>
<p><img src="getting_started_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] ggplot2_3.5.1 dplyr_1.1.4   gimap_0.1.0  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] tidyr_1.3.1        sass_0.4.9         utf8_1.2.4         generics_0.1.3    </span></span>
<span><span class="co">#&gt;  [5] xml2_1.3.6         stringi_1.8.4      hms_1.1.3          digest_0.6.36     </span></span>
<span><span class="co">#&gt;  [9] magrittr_2.0.3     RColorBrewer_1.1-3 evaluate_0.24.0    grid_4.4.1        </span></span>
<span><span class="co">#&gt; [13] timechange_0.3.0   fastmap_1.2.0      jsonlite_1.8.8     purrr_1.0.2       </span></span>
<span><span class="co">#&gt; [17] fansi_1.0.6        viridisLite_0.4.2  scales_1.3.0       textshaping_0.4.0 </span></span>
<span><span class="co">#&gt; [21] jquerylib_0.1.4    cli_3.6.3          crayon_1.5.3       rlang_1.1.4       </span></span>
<span><span class="co">#&gt; [25] bit64_4.0.5        munsell_0.5.1      withr_3.0.0        cachem_1.1.0      </span></span>
<span><span class="co">#&gt; [29] yaml_2.3.8         parallel_4.4.1     tools_4.4.1        tzdb_0.4.0        </span></span>
<span><span class="co">#&gt; [33] memoise_2.0.1      colorspace_2.1-0   kableExtra_1.4.0   vctrs_0.6.5       </span></span>
<span><span class="co">#&gt; [37] R6_2.5.1           lifecycle_1.0.4    lubridate_1.9.3    snakecase_0.11.1  </span></span>
<span><span class="co">#&gt; [41] stringr_1.5.1      bit_4.0.5          fs_1.6.4           vroom_1.6.5       </span></span>
<span><span class="co">#&gt; [45] ragg_1.3.2         janitor_2.2.0      pkgconfig_2.0.3    desc_1.4.3        </span></span>
<span><span class="co">#&gt; [49] pkgdown_2.0.9      pillar_1.9.0       bslib_0.7.0        gtable_0.3.5      </span></span>
<span><span class="co">#&gt; [53] glue_1.7.0         systemfonts_1.1.0  highr_0.11         xfun_0.45         </span></span>
<span><span class="co">#&gt; [57] tibble_3.2.1       tidyselect_1.2.1   rstudioapi_0.16.0  knitr_1.47        </span></span>
<span><span class="co">#&gt; [61] farver_2.1.2       htmltools_0.5.8.1  labeling_0.4.3     rmarkdown_2.27    </span></span>
<span><span class="co">#&gt; [65] svglite_2.1.3      readr_2.1.5        pheatmap_1.0.12    compiler_4.4.1</span></span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Candace Savonen, Howard Baek, Kate Isaac.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
